src/fomo.jl

```diff
@@ -0,0 +1,427 @@
+export fomo, FomoSolver, R2, R2Solver, tr, r2, R2og
+
+abstract type AbstractFomoMethod end
+
+struct tr   <: AbstractFomoMethod end
+struct r2   <: AbstractFomoMethod end
+struct R2og <: AbstractFomoMethod end
+
+"""
+    fomo(nlp; kwargs...)
+
+A First-Order with MOmentum (FOMO) model-based method for unconstrained optimization. Supports quadratic regularization and trust region methods.
+
+For advanced usage, first define a `FomoSolver` to preallocate the memory used in the algorithm, and then call `solve!`:
+
+    solver = FomoSolver(nlp)
+    solve!(solver, nlp; kwargs...)
+
+# Arguments
+- `nlp::AbstractNLPModel{T, V}` is the model to solve, see `NLPModels.jl`.
+
+# Keyword arguments 
+- `x::V = nlp.meta.x0`: the initial guess.
+- `atol::T = √eps(T)`: absolute tolerance.
+- `rtol::T = √eps(T)`: relative tolerance: algorithm stops when ‖∇f(xᵏ)‖ ≤ atol + rtol * ‖∇f(x⁰)‖.
+- `η1 = eps(T)^(1/4)`, `η2 = T(0.95)`: step acceptance parameters.
+- `γ1 = T(1/2)`, `γ2 = T(2)`: regularization update parameters.
+- `γ3 = T(1/2)` : momentum factor satβ update parameter in case of unsuccessful iteration.
+- `αmax = 1/eps(T)`: step parameter for fomo algorithm.
+- `max_eval::Int = -1`: maximum number of evaluation of the objective function.
+- `max_time::Float64 = 30.0`: maximum time limit in seconds.
+- `max_iter::Int = typemax(Int)`: maximum number of iterations.
+- `β = T(0.9) ∈ [0,1)` : target decay rate for the momentum.
+- `θ1 = T(0.1)` : momentum contribution parameter for convergence condition #1. [(1-satβ)∇f(xk) + satβ mk.∇f(xk)] ≥ θ1||∇f(xk)||², with mk memory of past gradient and satβ ∈ [0,β].
```

Comment:
What condition does this correspond to in your writeup? I don’t understand what `mk.∇f` mean. Use spacing to better differentiate terms: `(1 - satβ) ∇f(xk)`, etc.

Reply:
This would be the first inequality of Assumption 3 in the writeup, by injecting (3) and developing the left-hand side.
mk in the code comment correspond to $$b_{k-1}$$ in the writeup.

Reply:
could it be called `momentum`?

Reply:
I'm not quite sure how to call it. To me the momentum rather refers to the method, while `d` is the gradient-related direction and `m` is the "memory of past gradients", as it is sometimes refer to in the literature.

Reply:
Would `pastgrad` or `gradmemory` would be more suited ? I fear `momentum` could introduce confusion.

Reply:
I would think “momentum term” is the right name for that term, and is where the name of the method comes from.

Reply:
I'll rename the variable `momentum` then.